File: code/sorting/src/tree_sort/tree_sort.java
// Part of Cosmos by OpenGenus Foundation

public class Tree {

    public static void main(String[] args) {
        TreeSort myTree;
        myTree = new TreeSort(4);

        myTree.invert(new TreeSort(5));
        myTree.invert(new TreeSort(8));
        myTree.invert(new TreeSort(1));
        myTree.traverse(new KeyPrint());
    }
    
    static class TreeSort {
        private TreeSort left;
        private TreeSort right;
        private int key;

        public TreeSort(int key) {
            this.key = key;
        }

        private void invert(TreeSort node) {
            if (node.key < key) {
                if (left != null) left.invert(node);
                else {
                    left = node;
                }
            } else {
                if (right != null) right.invert(node);
                else {
                    right = node;
                }
            }
        }

        private void traverse(TreeVisitor visitor) {
            if (left != null)
                left.traverse(visitor);

            visitor.visit(this);

            if (right != null)
                right.traverse(visitor);
        }
    }

    interface TreeVisitor {
        void visit(TreeSort node);
    }

    static class KeyPrint implements TreeVisitor {
        public void visit(TreeSort node) {
            System.out.println(" " + node.key);
        }
    }
}

File: code/sorting/src/tree_sort/tree_sort.js
// Part of Cosmos by OpenGenus Foundation
var BinarySearchTree = function(value) {
  var instance = Object.create(BinarySearchTree.prototype);

  instance.value = value;
  // a BST where all values are higher than than the current value.
  instance.right = undefined;
  // a binary search tree (BST) where all values are lower than than the current value.
  instance.left = undefined;

  return instance;
};

BinarySearchTree.prototype.insert = function(value) {
  // accepts a value and places in the tree in the correct position.
  var node = BinarySearchTree(value);

  function recurse(bst) {
    if (bst.value > value && bst.left === undefined) {
      bst.left = node;
    } else if (bst.value > value) {
      recurse(bst.left);
    } else if (bst.value < value && bst.right === undefined) {
      bst.right = node;
    } else if (bst.value < value) {
      recurse(bst.right);
    }
  }

  recurse(this);
};

BinarySearchTree.prototype.contains = function(value) {
  var doesContain = false;
  //accepts a value and returns a boolean reflecting whether or not the value is contained in the tree.
  function recurse(bst) {
    if (bst.value === value) {
      doesContain = true;
    } else if (bst.left !== undefined && value < bst.value) {
      recurse(bst.left);
    } else if (bst.right !== undefined && value > bst.value) {
      recurse(bst.right);
    }
  }

  recurse(this);
  return doesContain;
};

BinarySearchTree.prototype.depthFirstLog = function(callback) {
  //accepts a callback and executes it on every value contained in the tree.
  function recurse(bst) {
    callback.call(bst, bst.value);

    if (bst.left !== undefined) {
      recurse(bst.left);
    }

    if (bst.right !== undefined) {
      recurse(bst.right);
    }
  }

  recurse(this);
};
